### PR TITLE
[three] Update object materials to match documentation

### PR DESCRIPTION
**Pull request by Methuselah96**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/mrdoob/three.js/issues/15015#issuecomment-427596484
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by Astrak** [comment]
Can't Line and Points accept `Material[]` as material, like Mesh?

**Comment by Methuselah96** [comment]
The documentation doesn't say it can. Can you point to where in the source code it allows for that?

File: types/aframe/test/aframe-io-tests.ts [hunk 1]
```diff
@@ -533,9 +533,10 @@ AFRAME.registerComponent('audioanalyser-waveform', {
 		rings.forEach(function transformRing(ring: THREE.Line, index: number) {
 			var normLevel;
 			normLevel = levels[RINGCOUNT - index - 1] + 0.01; // Avoid scaling by 0.
-			(ring.material as THREE.LineBasicMaterial).color.setHSL(colors[index], 1, normLevel);
-			ring.material.linewidth = normLevel * 3;
-			ring.material.opacity = normLevel;
+            const lineMaterial = ring.material as THREE.LineBasicMaterial;
```
Comment (efokschaner) [inline types/aframe/test/aframe-io-tests.ts]: Indentation looks off here.

**Comment by efokschaner** [comment]
In light of @mrdoobs comments this pr seems good, once the line material array question is resolved.

**Comment by Astrak** [comment]
I am not sure anyone knows what the situation is. What is the definition of a material type in ThreeJS: doc? name? no error? no rendering problem? Depending on the answer then the type implementation is different. Regarding the line and points material arrays, if we follow the logic of what does not throw an error, then they should be included.

Now, I believe we are trying to do things in lieu of ThreeJS, while they are having a problem. What is the definition of acceptable? According to naming convention and by reading the thread, the contributors have not considered whether it works or not. If we try it, some materials work, though they have their counterparts with the correct names for the appropriate object, so what are they for? 

But some don't have all the features of other objects, namely Line needs a shader with lineWidth on Mac, otherwise the width is always 1px, and Points need the PointSize otherwise the size is always 1px. Is it a problem? In the postprocessing, the overrideMaterial is used to apply a MeshDepthMaterial to all the objects in the scene, and while lineWidth is not used because it is Mac-only, at least the PointSize should be broken in the rendering with a MeshDepthMaterial. 

So, I would argue that the solution should rather be a PR in ThreeJS, that @types/three should later implement. There should be intermediate classes for Mesh/Line/Points so that the different material names have an effective restriction in the renderer, or the material setter and object constructor. The overrideMaterial seems a problem: it is mainly used to apply a MeshDepthMaterial in the posteffects. Maybe it can be deprecated in favor or a `scene.renderDepth` flag that instantiates and switches LineDepthMaterial and PointsDepthMaterial on the related objects when needed. If ones needs a custom shader as the overrideMaterial, it would be up to him to assign it to every objects instead.

**Comment by Astrak** [comment]
It seems that there is no clear definition of what things should be, and the ThreeJS thread contradicts itself. Should naming convention be followed? Doc? Script error handling? Rendering (is broken-rendering-without-error acceptable?)? Depending on what threshold the user should be limited to, the solution is different.

If the most lenient way is picked, still I can't see gl_PointSize in MeshDepthMaterial, or the sizeAttenuation property, meaning that the point rendering should be broken with the MeshDepthMaterial assigned in overrideMaterial. If so, then ThreeJS has a problem. Then, as suggested in the thread, why put a LineBasicMaterial on a Mesh? The shader will have the exact same effect, as for a PointsMaterial on a Mesh. These are the two cases where the necessary feature exactly match the needs. But we have MeshBasicMaterial for this, so why let a user do so... or why implement different shaders.

I believe that the correct solution should be a PR to ThreeJS, with intermediate classes for Line/Points/Meshes. The overrideMaterial feature shoould be discussed and possibly discarded, in favour of a `scene.renderDepth` flag that would switch to a LineDepthMaterial/PointsDepthMaterial/MeshDepthMaterial where needed, internally. Other overrideMaterial possibilities would be handled manually by the dev for each object.

**Comment by Methuselah96** [comment]
@Astrak I agree that it should be fixed in the three.js library. I have added arrays of materials until three.js is fixed to be more restrictive.

**Comment by Astrak** [comment]
@Methuselah96 Unsure to understand the backdrop of the PR. What has concluded to reversing back? The related discussions have been looking inconclusive to me.

**Comment by Methuselah96** [comment]
@Astrak My conclusion is that the three.js API does not make any guarantees as to which materials can be applied to which objects and therefore we shouldn't restrict the user in that regard. I think the fact that you might want to apply a `MeshDepthMaterial` to a `Points` or `Line` is a perfect example of this. I think the impetus is now on three.js to fix its API to more clearly define what materials should be applied and which ones shouldn't and until that is fixed the types should reflect the uncertainty of the API. I completely agree with your diagnosis of the changes that need to be made to three.js in order to fix the API and I think those changes need to happen on the three.js library before the types can reflect those changes, which is why I'm making this PR to make the types reflect the API.

**Comment by Astrak** [comment]
> the types should reflect the uncertainty of the API

Fair enough :smile:

**Review by Astrak — APPROVED** [review]

**Pull request merged**